Bluetooth: Add GATT Caching support

Vudentz opened the pull request:
This implements GATT Caching support introduced in Bluetooth 5.1.

Vudentz committed:
Bluetooth: Add definitions for GATT Caching

This adds the definitions for the GATT Caching features.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Implement Client Supported Features

The Client Supported Features characteristic is used by the client to
inform the server which features are supported by the client.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Implement Database Hash

This implements the Database Hash characteristic which generates a hash
with the contents of certain attributes. The generation of hash is
usually offloaded to the systemwq using a delayed work so that when
application register multiple services only one hash needs to be
calculated.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: Kconfig: Add BT_GATT_CACHING

This adds BT_GATT_CACHING option which can be used to enable support
for Client Supported Features and Database Hash characteristics.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Implement Robust Caching

This implement Robust Caching which is mandatory when Database Hash and
Service changed Characteristics are supported.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Persist Client Features config only when paired

BLUETOOTH CORE SPECIFICATION Version 5.1 | Vol 3, Part G page 2405:

  'For clients with a trusted relationship, the characteristic value
  shall be persistent across connections. For clients without a
  trusted relationship the characteristic value shall be set to the
  default value at each connection.'

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: Kconfig: Add menu for ATT and GATT options

This move Kconfig options for ATT and GATT to a dedicated menu.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

aescolar commented:
Thanks for that @Vudentz :).
Note that codecov is reporting that of the new ~140 lines of code in gatt.c which could have been covered, only 43 are in our CI tests.
Would you think adding a test to try to cover a bit more of it would be worthwhile?

Vudentz commented:
> Thanks for that @Vudentz :).
> Note that codecov is reporting that of the new ~140 lines of code in gatt.c which could have been covered, only 43 are covered in our CI tests.
> https://codecov.io/gh/zephyrproject-rtos/zephyr/pull/13070/src/subsys/bluetooth/host/gatt.c?before=subsys/bluetooth/host/gatt.c
> Would you think adding a test to try to cover a bit more of it would be worthwhile?

Sure, though perhaps we should start with auto-pts, for covering all this code we would need a client which implements GATT caching as well though that is normally not used by systems running zephyr since we don't have the luxury of caching the whole database of the remote devices are connecting to.

aescolar commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -354,6 +325,49 @@ config BT_L2CAP_DYNAMIC_CHANNEL
+	select TINYCRYPT_AES_CMAC
+	help
+	  This option enables support for GATT Caching. When enabled the stack
+	  witll register Client Supported Features and Database Hash
```
silly thing: wi*t*ll

thoh-ot commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -267,6 +275,129 @@ static ssize_t cf_write(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+		}
+
+		if (tc_cmac_update(&state->state,
+				   (uint8_t *)&sys_cpu_to_le16(u16->val),
```
See earlier comment regarding sys_cpu_to_le16.

thoh-ot commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -267,6 +275,129 @@ static ssize_t cf_write(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+	case 0x2904: /* GATT Characteristic Presentation Format */
+	case 0x2905: /* GATT Characteristic Aggregated Format */
+		if (tc_cmac_update(&state->state,
+				   (uint8_t *)&sys_cpu_to_le16(attr->handle),
```
See earlier comment regarding sys_cpu_to_le16.

thoh-ot commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -267,6 +275,129 @@ static ssize_t cf_write(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+		}
+
+		if (tc_cmac_update(&state->state,
+				   (uint8_t *)&sys_cpu_to_le16(u16->val),
```
See earlier comment regarding sys_cpu_to_le16.

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -267,6 +275,129 @@ static ssize_t cf_write(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+	case 0x2803: /* GATT Characteristic */
+	case 0x2900: /* GATT Characteristic Extended Properties */
+		if (tc_cmac_update(&state->state,
+				   (uint8_t *)&sys_cpu_to_le16(attr->handle),
```
Doesn't __bswap_16 return a new value, I would expect the compiler to just store it for me in a unnamed variable, but perhaps using a variable here is clearer. Btw, we should probably have a big endian board to have the CI check these details automatically.

thoh-ot replied:
As both  sys_cpu_to_le16 and __bswap_16 are macros what will end up happening is that:

`(uint8_t *)&sys_cpu_to_le16(attr->handle)`

becomes:

- LE: `(uint8_t*)&(attr->handle)`
- BE: `(uint8_t*)&((int) ((((attr->handle) >> 8) & 0xff) | (((attr->handle) & 0xff) << 8)))` and gcc gives the error: _lvalue required as unary '&' operand_

I agree regarding having a big-endian board in CI (especially since Oticon is using a big-endian cpu), but this issue could actually just be checked by a big-endian compilation.

Vudentz committed:
Bluetooth: GATT: Store Client features

Client features shall be persisted across connections.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
@thoh-ot Does that looks better now?
@jhedberg Any chance to review this one before the freeze?

aescolar commented:
Note: The coverage report massive diff would seem to be due to codecov not being able to find the report from the master which this commit was merged against in CI, and quite confusedly picking another one as reference from quite back in time

Vudentz commented:
> Note: The coverage report massive diff would seem to be due to codecov not being able to find the report from the master which this commit was merged against in CI, and quite confusedly picking another one as reference from quite back in time. Meaning, you can ignore it

I just rebased lets see if that takes care of the massive diff.

aescolar commented:
> I just rebased lets see if
I don't think it is due to anything you have done (or could do) but let's see :), maybe when this new coverage run has finished, codecov has already the result from Jukka's new commits

aescolar commented:
@Vudentz : The coverage report is sensible again :) (there a lot of merges to master and load in CI..)

jhedberg commented:
Have you had the chance to run this against the PTS or any other implementation?

thoh-ot commented:
@Vudentz Can't seem to find any code relating to BLUETOOTH CORE SPECIFICATION Version 5.1 | Vol 3, Part G page 2350: "If a change-unaware client sends an ATT command, the server shall ignore it. Except for the Handle Value indication, the server shall not send notifications and indications to such a client until it becomes change-aware."? Maybe i'm not looking hard enough.

Vudentz commented:
> Have you had the chance to run this against the PTS or any other implementation?

I have been running this against BlueZ, not sure if we will have time until tomorrow to validate with PTS but I suppose we can send fixes if we find out there is something which needs fixing.

Vudentz commented:
> @Vudentz Can't seem to find any code relating to BLUETOOTH CORE SPECIFICATION Version 5.1 | Vol 3, Part G page 2350: "If a change-unaware client sends an ATT command, the server shall ignore it. Except for the Handle Value indication, the server shall not send notifications and indications to such a client until it becomes change-aware."? Maybe i'm not looking hard enough.

The ATT commands would fail as they check for bt_gatt_change_aware, on the other hand I haven't enforce anything on notification and indication since that may break stuff that have been working just fine, also note it is a grey area what the server would do with those since they can be either queued up or discarded and none of those seems to be the right thing when the server knows their handles haven't been changed.

thoh-ot commented:
How do you interpret BLUETOOTH CORE SPECIFICATION Version 5.1 | Vol 3, Part G page 2349:
"If a client that has indicated support for robust caching (by setting the Robust
Caching bit in the Client Supported Features characteristic) is change-unaware
then, when that client requests an operation at any Attribute Handle or list of
Attribute Handles by sending an ATT request, the server shall not perform the
operation but instead shall send a response with the error code set to
Database Out Of Sync." w.r.t reading the Database Hash? Should it be read or not? It seems like in figure 2.6 it's read as one on the response arrows contains the text "ATT Response (Database Out Of Sync OR Hash Read)".

But if i look at GATT.TS.5.1.0 4.8.8 GATT/SR/GAS/BV-05-C [Handling Client Requests on Unsynchronized Database]; test procedure step 14:
"14. The Lower Tester executes the GATT Characteristic Value Read sub-procedure on the Database Hash characteristic and expects the IUT to return the Database Out of Sync error.", it seems it should not be read...

thoh-ot commented:
> > @Vudentz Can't seem to find any code relating to BLUETOOTH CORE SPECIFICATION Version 5.1 | Vol 3, Part G page 2350: "If a change-unaware client sends an ATT command, the server shall ignore it. Except for the Handle Value indication, the server shall not send notifications and indications to such a client until it becomes change-aware."? Maybe i'm not looking hard enough.
> 
> The ATT commands would fail as they check for bt_gatt_change_aware, on the other hand I haven't enforce anything on notification and indication since that may break stuff that have been working just fine, also note it is a grey area what the server would do with those since they can be either queued up or discarded and none of those seems to be the right thing when the server knows their handles haven't been changed.

Ah yes the req check in att_write_rsp "!bt_gatt_change_aware(conn, req ? true : false)". Didn't catch that first time around.

Vudentz commented:
> How do you interpret BLUETOOTH CORE SPECIFICATION Version 5.1 | Vol 3, Part G page 2349:
> "If a client that has indicated support for robust caching (by setting the Robust
> Caching bit in the Client Supported Features characteristic) is change-unaware
> then, when that client requests an operation at any Attribute Handle or list of
> Attribute Handles by sending an ATT request, the server shall not perform the
> operation but instead shall send a response with the error code set to
> Database Out Of Sync." w.r.t reading the Database Hash? Should it be read or not? It seems like in figure 2.6 it's read as one on the response arrows contains the text "ATT Response (Database Out Of Sync OR Hash Read)".

In case the client is change-unware it will always return out-of-sync error except for reading DB Hash, note that this only applies to procedures operating on handles directly. There are conditions to become change-aware such as having a request be sent' twice.
 
> But if i look at GATT.TS.5.1.0 4.8.8 GATT/SR/GAS/BV-05-C [Handling Client Requests on Unsynchronized Database]; test procedure step 14:
> "14. The Lower Tester executes the GATT Characteristic Value Read sub-procedure on the Database Hash characteristic and expects the IUT to return the Database Out of Sync error.", it seems it should not be read...

Yep, that is because it using the Read procedure not the Read by Type, this is because some servers may change the whole range 0x0001-0xffff which means the DB Hash would also be affected.

thoh-ot commented:
I'm not that well versed in in the host stuff about persistence and settings. But the other parts seems fine to me.

Vudentz commented:
> I'm not that well versed in in the host stuff about persistence and settings. But the other parts seems fine to me.

Ive pushed a new version which has a comment regarding the handling of Write Command on bt_gatt_change_aware, Im also thinking on return an error in bt_gatt_notify/bt_gatt_indicate if we detect the client is change-unaware though in case of sending to any connection (conn == NULL) we might not be able to do that.

Vudentz committed:
Bluetooth: GATT: Add option to enforce client change-unware state

This adds BT_GATT_ENFORCE_CHANGE_UNAWARE option which when enable
returns -EAGAIN when notifying or indicating if the client is
change-unware to conform with following statement on the spec:

  'BLUETOOTH CORE SPECIFICATION Version 5.1 | Vol 3, Part G page 2350:
   Except for the Handle Value indication, the  server shall not send
   notifications and indications to such a client until it becomes
   change-aware.'

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
@jhedberg Lets try to get this merged since all of the comments seems to be resolved.
